openedx/core/djangoapps/oauth_dispatch/migrations/0003_application_data.py

```diff
@@ -0,0 +1,38 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.13 on 2018-06-05 13:19
+from __future__ import unicode_literals
+
+from django.conf import settings
+from django.db import migrations
+
+
+def migrate_application_data(apps, schema_editor):
+    """
+    Migrate existing DOT Application models to new ScopedApplication models.
+    """
+    Application = apps.get_model(settings.OAUTH2_PROVIDER_APPLICATION_MODEL)
+    ScopedApplication = apps.get_model('oauth_dispatch', 'ScopedApplication')
+
+    for application in Application.objects.all():
+        scoped_application = ScopedApplication(
+            client_id=application.client_id,
+            user=application.user,
+            redirect_uris=application.redirect_uris,
+            client_type=application.client_type,
+            authorization_grant_type=application.authorization_grant_type,
+            client_secret=application.client_secret,
+            name=application.name,
+            skip_authorization=application.skip_authorization,
+        )
+        scoped_application.save()
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('oauth_dispatch', '0002_scopedapplication_scopedapplicationorganization'),
+    ]
+
+    operations = [
+        migrations.RunPython(migrate_application_data),
```

Comment:
Questions about this migration:

1. Can we pass `reverse_code=RunPython.noop` to be explicit that there's no rollback of this migration?
2. What happens if the migration is run, rolled back, and then re-run?  Will `migrate_application_data` handle that?
3. What if `migrate_application_data` fails in the middle?  Will re-running the migration coalesce any issues?
4. How will we prevent or detect that no new `Application` entries get created in between this PR and the one that changes `settings.OAUTH2_PROVIDER_APPLICATION_MODEL`?  Should we create another small PR that changes the value of `settings.OAUTH2_PROVIDER_APPLICATION_MODEL`?  Should we move this migration to a management command that runs again (just in case)?

#2 and #3 are all issues that have actually happened in production, so not just theoretical.

Reply:
1. Done.
2+3. Switched to `get_or_create` instead of `save`.
4. We should accept that risk. We can do a manual check after we switch to the new application model to verify that the two application tables are in sync.

Reply:
@feanil FYI on this migration strategy.

Reply:
Adding:
5. The "id" field in the new and old Application models need to match exactly in order to support all previously created tokens/grants/etc.  Those tables have foreign keys to the Application table via the "id" field.